### PR TITLE
Refactor TOSA backend preprocess to use visitor pattern

### PR DESCRIPTION
**Pull request by Jerry-Ge**
No lowering logic change. Only a very large refactor patch to make the current code more modular for easier future maintenance and development.

### CODE REVIEW

File: backends/arm/arm_vela.py [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+def vela_compile(tosa_graph):
+    with tempfile.TemporaryDirectory() as tmpdir:
+        tosaname = "out.tosa"
+        flatbuffer = tosa_graph.serialize()
+        with open(os.path.join(tmpdir, tosaname), "wb") as f:
+            f.write(flatbuffer)
+
+        # invoke vela
+        vela_command = (
+            f"cd {tmpdir}; vela --accelerator-config ethos-u55-128 {tosaname}"
```
Comment (digantdesai) [inline backends/arm/arm_vela.py:38]: nit take accelerator config name as an optional arg which defaults to this?

File: backends/arm/operators/node_visitor.py [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+    """
+
+    def __init__(self, exported_program: ExportedProgram):
+        self._exported_program = exported_program or None
+
+    def define_node(
+        self,
+        node: torch.fx.Node,
+        tosa_graph: ts.TosaSerializer,
+        inputs,
```
Comment (digantdesai) [inline backends/arm/operators/node_visitor.py:25]: type annotations?

File: backends/arm/test/test_models.py [hunk 3]
```diff
@@ -431,7 +446,7 @@ def __init__(self):
         def forward(self, x):
             return self.softmax(x)
 
-    # @register_test
+    @register_test
```
Comment (digantdesai) [inline backends/arm/test/test_models.py:449]: why these changes with refactor PR?

File: backends/arm/test/test_models.py [hunk 4]
```diff
@@ -41,6 +41,21 @@ class TorchBuilder:
     def __init__(self):
         pass
 
+    @register_test
```
Comment (digantdesai) [inline backends/arm/test/test_models.py:44]: why these changes with refactor PR?

**Review by digantdesai — COMMENTED** [review]
So clean!! :) Left some comments but lgtm.

nit - seems like there are some changes which are not purely refactor, perhaps move them to another PR?

**Comment by cccclai** [comment]
This is great! I have no idea how you manage to read the previous code....

File: backends/arm/test/test_models.py (hunk 4 above)
Comment (Jerry-Ge) [reply backends/arm/test/test_models.py]: reverted.

File: backends/arm/test/test_models.py (hunk 3 above)
Comment (Jerry-Ge) [reply backends/arm/test/test_models.py]: reverted.

File: backends/arm/operators/node_visitor.py (hunk 2 above)
Comment (Jerry-Ge) [reply backends/arm/operators/node_visitor.py]: done.

File: backends/arm/arm_vela.py (hunk 1 above)
Comment (Jerry-Ge) [reply backends/arm/arm_vela.py:38]: since this is not refactoring related, let's put it another tiny patch : )

**Comment by Jerry-Ge** [comment]
> This is great! I have no idea how you manage to read the previous code....

I don't think you can 🙂

**Comment by Jerry-Ge** [comment]
the lint-runner failure should not be related to this patch
```
+ lintrunner --force-color --all-files --tee-json=lint.json


>>> Lint for sdk/size_analysis_tool/size_analysis_tool.py:

  Warning (FLAKE8) F401
    'executorch.sdk.etrecord._etrecord.ETRecordReservedFileNames' imported
    but unused
    See https://www.flake8rules.com/rules/F401.html.

```

File: backends/arm/tosa_utils.py [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+    return input_transposed
+
+
+def getNodeArgs(node):
+    return [TosaArg(arg) for arg in node.args]
+
+
+# Helper function to do broadcasting
+# Ref: https://www.mlplatform.org/tosa/tosa_spec.html#_broadcasting
+def broadcastShapes(shape1, shape2):
```
[the remark this answers is not in the document]
Comment (Jerry-Ge) [reply backends/arm/tosa_utils.py:104]: yes: https://www.mlplatform.org/tosa/tosa_spec.html#_broadcasting

**Comment by digantdesai** [comment]
> the lint-runner failure should not be related to this patch
> 
> ```
> + lintrunner --force-color --all-files --tee-json=lint.json
> 
> 
> >>> Lint for sdk/size_analysis_tool/size_analysis_tool.py:
> 
>   Warning (FLAKE8) F401
>     'executorch.sdk.etrecord._etrecord.ETRecordReservedFileNames' imported
>     but unused
>     See https://www.flake8rules.com/rules/F401.html.
> ```

yeah should be fixed soon, feel free to ignore

**Review by digantdesai — APPROVED** [review]
I am assuming you ran e2e and also setup.sh and run.sh from examples/arm dir?
